admin/cli.py

```diff
@@ -226,46 +267,95 @@ def stop(c, tunnel_id):
     # being lazy and overzealous at the same time - we'll just garbage-college its resources.
     gc(c)
 
+
 @task
 def connect(c, tunnel_id, command="/bin/bash", pty=True):
     """ Connect to a remote device, identified by a tunnel. """
-    # Care should be exercised here; we're taking data from a remote source and using it to
-    # run shell commands. Validate every last bit of data.
-    t = get_tunnel(tunnel_id)
-    assert TunnelState(t['state']) == TunnelState.running, "Device has not yet connected"
-    dip = device_ip(IPv4Network(t['network'])).ip
-    assert t['support_user'].isalnum()
-    assert t['support_user'].isascii()
-    support_user = t['support_user']
+    device = connect_tunnel(c, tunnel_id)
+    # and finally execute a shell
+    device.sudo(command, pty=pty)
 
-    # set up local
-    c.run("gcloud compute config-ssh", hide="both")
-    user_from_oslogin = c.run("gcloud compute os-login describe-profile --format=json", hide="both")
-    ts_user = json.loads(user_from_oslogin.stdout)['posixAccounts'][0]['username']
 
-    # set up connection to bastion
-    ts = Connection(
-        host = str(get_ts_instance_public_ip(tunnel_id)),
-        user = ts_user,
-        connect_kwargs={"auth_timeout": 120} # long for 2FA
-    )
+def is_remote_directory(remote_device, remote_path):
+    result = remote_device.run(f"if [ -d '{remote_path}' ]; then echo 'directory'; else echo 'file'; fi", hide=True).stdout.strip()
+    return result == 'directory'
 
-    # grab the ssh private key on the tunnel server
-    ssh_privkey = ts.run(f"sudo cat {SSH_KEYFILE_PATH}", hide="both")
-    assert ssh_privkey
 
-    # set up connection to destination device
-    device = Connection(
-        host=str(dip),
-        user=support_user,
-        gateway=ts,
-        connect_kwargs = {
-            "pkey": Ed25519Key.from_private_key(io.StringIO(ssh_privkey.stdout)),
-        }
-    )
+def send_file(remote_device, local_file, remote_dir):
+    filename = path.basename(local_file)
+    remote_path = path.join(remote_dir, filename).replace("./", "")  # path.join adds an unneccessary ./ when combining things with their own directory, making the prints down the line look odd
+
+    # Support user lacks permissions to send file to just any directory, scrape the filename and send to an intermediary and then sudo mv it to the proper location
+    print(f"Uploading {local_file} to {remote_path}...")
+    temp_remote = f"/tmp/{filename}"
+    remote_device.put(local=local_file, remote=temp_remote, preserve_mode=True)
+    remote_device.sudo(f"mv {temp_remote} {remote_path}", pty=False)
+    print(f"File successfully uploaded to {remote_path}")
```

Comment:
I wanted this section to also have a `remote_device.sudo('mkdir -p /tmp/support')` and use the support folder instead just in case there happens to be stuff directly in `/tmp` that would be replaced, but that particular new directory consistently didn't let me send files to it for a permissions error
Just `/tmp` is fine